lib/Parse/SyntaxRegexFallbackLexing.cpp

```diff
@@ -0,0 +1,147 @@
+//===--- SyntaxRegexFallbackLexing.cpp ------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Parse/SyntaxRegexFallbackLexing.h"
+#include "swift/AST/DiagnosticEngine.h"
+#include "swift/AST/DiagnosticsParse.h"
+#include "swift/Parse/Lexer.h"
+#include "swift/Parse/RegexParserBridging.h"
+#include <mutex>
+
+using namespace swift;
+
+template <typename... DiagArgTypes, typename... ArgTypes>
+static void diagnose(BridgedOptionalDiagnosticEngine bridgedDiag,
+                     const char *ptr, Diag<DiagArgTypes...> DiagID,
+                     ArgTypes &&...Args) {
+  if (auto *Diag = static_cast<DiagnosticEngine *>(bridgedDiag.object)) {
+    Diag->diagnose(SourceLoc(llvm::SMLoc::getFromPointer(ptr)), DiagID,
+                   std::forward<ArgTypes>(Args)...);
+  }
+}
+
+bool syntaxparse_lexRegexLiteral(
+    const char **InputPtr, const char *BufferEnd, bool MustBeRegex,
+    BridgedOptionalDiagnosticEngine BridgedDiagEngine) {
+
+  const char *Ptr = *InputPtr;
+
+  // Count leading '#'.
+  while (*Ptr == '#') {
+    ++Ptr;
+  }
+  if (*Ptr != '/') {
+    // This wasn't a regex literal.
+    return true;
+  }
+
+  unsigned customDelimiterLen = Ptr - *InputPtr;
+
+  ++Ptr;
+
+  // If the delimiter allows multi-line, try skipping over any whitespace to a
+  // newline character. If we can do that, we enter multi-line mode.
+  bool allowsMultiline = customDelimiterLen != 0;
+  const char *firstNewline = nullptr;
+  if (allowsMultiline) {
+    while (Ptr != BufferEnd) {
+      switch (*Ptr) {
+      case ' ':
+      case '\t':
+        ++Ptr;
+        continue;
+      case '\r':
+      case '\n':
+        firstNewline = Ptr;
+        break;
+      default:
+        break;
+      }
+      break;
+    }
+  }
+
+  while (true) {
+    switch (*Ptr++) {
+    case '\r':
+    case '\n':
+      if (firstNewline == nullptr) {
+        // not a multiline literal
+        diagnose(BridgedDiagEngine, Ptr, diag::lex_regex_literal_unterminated);
+        *InputPtr = Ptr - 1;
+        return false;
+      }
+      break;
+    case '\\':
+      if (Ptr != BufferEnd) {
+        if (validateUTF8CharacterAndAdvance(Ptr, BufferEnd) == ~0U)
+          diagnose(BridgedDiagEngine, Ptr, diag::lex_invalid_utf8);
+      }
```

Comment:
Could this skip over a newline if you had something like the following?

```swift
let r = /\
/
```

Reply:
Good catch. That could indeed happen. Just to be clear though, the following is supported, right

```swift
let r = #/
\
/#
```

Reply:
Yup, that's correct

Reply:
Interesting